### PR TITLE
ISSD-2694 [PRM] Create a Client Extension Spring Boot for Account Sync

### PR DESCRIPTION
**Pull request by thaynam**

### CODE REVIEW

**Comment by thaynam** [comment]
ci:test:sf

**Comment by thaynam** [comment]
ci:test:stable

**Review by jairmedeiros — CHANGES_REQUESTED** [review]

**Review by jairmedeiros — CHANGES_REQUESTED** [review]
Details

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-partner-portal/src/main/resources/site-initializer/object-definitions/account-sf.json
Comment (jairmedeiros) [inline modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-partner-portal/src/main/resources/site-initializer/object-definitions/account-sf.json:1]: Rename to proxy-accounts.json

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-partner-portal/src/main/resources/site-initializer/object-definitions/account-sf.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+			"required": false,
+			"type": "String"
+		}
+	],
+	"panelCategoryKey": "control_panel.object",
+	"pluralLabel": {
+		"en_US": "Accounts SF"
+	},
+	"scope": "company",
+	"storageType": "salesforce"
```
Comment (jairmedeiros) [inline modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-partner-portal/src/main/resources/site-initializer/object-definitions/account-sf.json:27]:
```suggestion
	"externalReferenceCode": "Account",
	"label": {
		"en_US": "Proxy Account"
	},
	"name": "ProxyAccount",
	"objectFields": [
		{
			"DBType": "String",
			"businessType": "Text",
			"externalReferenceCode": "Name",
			"indexed": true,
			"indexedAsKeyword": false,
			"label": {
				"en_US": "Name"
			},
			"name": "name",
			"required": false,
			"type": "String"
		}
	],
	"panelCategoryKey": "control_panel.object",
	"pluralLabel": {
		"en_US": "Proxy Accounts"
	},
	"scope": "company",
	"storageType": "salesforce"
```

File: workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+					partnerLevelJSONObject.getJSONObject("partnerLevelType");
+
+				String partnerLevelType = partnerLevelTypeJSONObject.optString(
+					"key");
+
+				String partnerLevelERC = partnerLevelJSONObject.getString(
+					"externalReferenceCode");
+
+				partnerLevelERCs.put(partnerLevelType, partnerLevelERC);
+			}
```
Comment (jairmedeiros) [inline workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java:426]:
```suggestion
			for (int i = 0; i < partnerLevelJSONArray.length(); i++) {
				JSONObject partnerLevelJSONObject =
					partnerLevelJSONArray.getJSONObject(i);

				JSONObject partnerLevelTypeJSONObject =
					partnerLevelJSONObject.getJSONObject("partnerLevelType");

				String partnerLevelType = partnerLevelTypeJSONObject.optString(
					"key");

				String partnerLevelERC = partnerLevelJSONObject.getString(
					"externalReferenceCode");

				partnerLevelERCs.put(partnerLevelType, partnerLevelERC);
			}
```

File: workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+					if (accountsfJSONObject.has("partnerLevelType")) {
+						JSONObject partnerLevelTypeJSONObject =
+							accountsfJSONObject.getJSONObject(
+								"partnerLevelType");
+
+						String partnerLevelType =
+							partnerLevelTypeJSONObject.getString("key");
+
+						Map<String, String> partnerLevelERCs =
+							_fetchPartnerLevelERCs();
```
Comment (jairmedeiros) [inline workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java:147]:
```suggestion
							_getPartnerLevelExternalReferenceCodes();
```

File: workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+
+					if (accountsfJSONObject.has("partnerLevelType")) {
+						JSONObject partnerLevelTypeJSONObject =
+							accountsfJSONObject.getJSONObject(
+								"partnerLevelType");
+
+						String partnerLevelType =
+							partnerLevelTypeJSONObject.getString("key");
+
+						Map<String, String> partnerLevelERCs =
```
Comment (jairmedeiros) [inline workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java:146]:
```suggestion
						Map<String, String> partnerLevelExternalReferenceCodes =
```

File: workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+
+						sb.append("Account: ");
+						sb.append(accountName);
+						sb.append(" (");
+						sb.append(salesforceAccountKey);
+						sb.append(") was UPDATED: ");
+						sb.append(partnerAccountJSONObject);
+
+						_log.info(sb.toString());
+					}
```
Comment (jairmedeiros) [inline workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java:193]: Improve this log.

File: workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+				).put(
+					"name", accountName
+				);
+
+				if (partnerCountry != null) {
+					partnerAccountJSONObject.put(
+						"partnerCountry", partnerCountry);
+				}
+
+				String accountRegion = "";
```
Comment (jairmedeiros) [inline workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java:129]:
```suggestion
				String accountRegion = null;
```

File: workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+				}
+
+				try {
+					_put(
+						partnerAccountJSONObject.toString(),
+						StringBundler.concat(
+							"/o/headless-admin-user/v1.0/accounts",
+							"/by-external-reference-code/",
+							salesforceAccountKey));
+
```
Comment (jairmedeiros) [inline workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java:181]:
```suggestion
					blabla = _put(
						partnerAccountJSONObject.toString(),
						StringBundler.concat(
							"/o/headless-admin-user/v1.0/accounts",
							"/by-external-reference-code/",
							salesforceAccountKey));

```

File: workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+	private Map<String, Long> _fetchRegionsIDs() {
+		Map<String, Long> regionsIDs = new HashMap<>();
+
+		JSONObject globalOrganizationResponseJSONObject = _get(
+			uriBuilder -> uriBuilder.path(
+				"/o/headless-admin-user/v1.0/organizations" +
+					"/by-external-reference-code/PRM-ORG-GLOBAL"
+			).queryParam(
+				"pageSize", "-1"
+			).build());
```
Comment (jairmedeiros) [inline workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java:441]:
```suggestion
			).build());
```

File: workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java [hunk 9]
```diff
@@ -0,0 +1,631 @@
...
+
+			for (int o = 0; o < organizationsJSONArray.length(); o++) {
+				JSONObject organizationJSONObject =
+					organizationsJSONArray.getJSONObject(o);
+
+				String organizationName = organizationJSONObject.getString(
+					"name");
+				Long organizationID = organizationJSONObject.getLong("id");
+
+				regionsIDs.put(organizationName, organizationID);
```
Comment (jairmedeiros) [inline workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java:466]:
```suggestion
			for (int i = 0; i < organizationsJSONArray.length(); i++) {
				JSONObject organizationJSONObject =
					organizationsJSONArray.getJSONObject(i);

				String organizationName = organizationJSONObject.getString(
					"name");
				Long organizationID = organizationJSONObject.getLong("id");

				regionsIDs.put(organizationName, organizationID);
```

File: workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java [hunk 10]
```diff
@@ -0,0 +1,631 @@
...
+					"externalReferenceCode");
+
+				partnerLevelERCs.put(partnerLevelType, partnerLevelERC);
+			}
+		}
+
+		return partnerLevelERCs;
+	}
+
+	private Map<String, Long> _fetchRegionsIDs() {
```
Comment (jairmedeiros) [inline workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java:432]:
```suggestion
	private Map<String, Long> _getRegionsIds() {
```

File: workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java (hunk 5 above)
Comment (thaynam) [reply workspaces/liferay-partner-workspace/client-extensions/liferay-partner-etc-spring-boot/src/main/java/com/liferay/partner/QueueListener.java:193]: Any sugestion?

**Comment by thaynam** [comment]
ci:test:sf

**Pull request closed without merging**